Lib/test/test_importlib/identity.py

```diff
@@ -0,0 +1,521 @@
+    return result
+
+
+@st.composite
+def identity_entries(draw, state, entry_strategy):
+    """
+    Generate a list of processed identity entries using
+    ``entry_strategy``.
+    """
+    sample = draw(st.lists(entry_strategy(state), min_size=0, max_size=10))
+    for entry in sample[:-1]:
+        entry.append(Token(identity_entries, ", ", True))
+    unbalance(sample)
+    lstrip(sample)
+    return process(sample)
+
+
+def process(entries):
+    """
+    Process a list of entries into structured and unstructured
+    formats. Filter empty (``None``) entries from the structured
+    format. Convert token categories to strings for simpler
+    visualization.
+    """
+    text = ""
+    data = set()
+
+    for entry in entries:
+        entry_text = entry.as_text
+        entry_data = entry.as_data
+        for token in entry:
+            token.category = token.category.__name__
+        text += entry_text
+        if all(elem is None for elem in entry_data):
+            continue
+        data.add(entry_data)
+
+    return (entries, data, text)
+
+
+def lstrip(entries):
+    """
+    The PackageMetadata implementation removes leading whitespace
+    from the first entry, so we do the same. This should not alter
+    the parsing of the entry.
+    """
+    entry = entries[0] if entries else []
+    strip = False
+    for token in entry:
+        value = token.value.lstrip()
+        strip = strip or token.value != value
+        token.value = value
+        if token.value:
+            break
+    return strip
+
+
+def unbalance_indexes(entries):
+    """Return a list of valid replacement indices for quote characters."""
+    index_candidates = []
+    type_candidates = [
+        ident_addr_domain_other,
+        ident_addr_local_other,
+        ident_name_other,
+        name_value_other,
+    ]
+    stop = False
+    i = len(entries) - 1
+    while i >= 0 and not stop:
+        j = len(entries[i]) - 1
+        while j >= 0 and not (stop := entries[i][j].category is quote):
+            token = entries[i][j]
+            if token.category not in type_candidates:
+                repls = ()
+            elif token.category is ident_name_other:
+                final = entries[i][j + 1] is ident_name_only
+                repls = get_name_qchar_idxs(token.value, not final)
+            else:
+                repls = range(len(token.value))
+            for k in repls:
+                index_candidates.append((i, j, k))
+            j -= 1
+        i -= 1
+    return index_candidates
+
+
+@st.composite
+def unbalance(draw, entries):
+    """
+    Unbalance quote characters in the list of entries without altering
+    the parsing of the list.
+    """
+    qchrs = draw(st.sets(st.sampled_from("\"'")))
+    index_candidates = unbalance_indexes(entries)
+    if not qchrs or not index_candidates:
+        return False
+    for qchr in qchrs:
+        idx_entry, idx_token, idx_char = draw(st.sampled_from(index_candidates))
```

Comment:
I'd use `sorted(qchrs)`: Hypothesis works best if replaying a choice sequence is always valid and gives the same result, but nondeterministic set iteration order could break that here.